File: rks/src/utils/compiler.rs
#![allow(non_snake_case)]

pub mod compiler {
    use std::collections::HashMap;
	use to_binary::BinaryString;

    use crate::utils::error::error::Error;

    pub struct Compiler<'a> {
		instructions: HashMap<&'a str, (&'a str, u8)>,
		registers: HashMap<&'a str, i8>,
		line: String,
		error: Error
	}

	impl Compiler<'_> {
		pub fn new(line: &String, error: Error) -> Self {
			Self { 
				instructions: HashMap::from(
					[
						("NOP", ("000000", 0)),
						("LDA", ("000001", 1)),
						("LDB", ("000010", 1)),
						("LDC", ("000011", 1)),
						("LDD", ("000100", 1)),
						("SWP", ("000101", 2)),
						("MOV", ("000110", 2)),
						("STA", ("000111", 1)),
						("STB", ("001000", 1)),
						("STC", ("001001", 1)),
						("STD", ("001010", 1)),
						("LRA", ("001011", 1)),
						("LRB", ("001100", 1)),
						("LRC", ("001101", 1)),
						("LRD", ("001110", 1)),
						("PSH", ("001111", 1)),
						("PUL", ("010000", 1)),
						("ISP", ("010001", 1)),
						("DSP", ("010010", 1)),
						("ADD", ("010011", 2)),
						("SUB", ("010100", 2)),
						("MUL", ("010101", 2)),
						("DIV", ("010110", 2)),
						("AND", ("010111", 2)),
						("OR", ("011000", 2)),
						("NOR", ("011001", 2)),
						("XOR", ("011010", 2)),
						("NOT", ("011011", 1)),
						("INC", ("011100", 1)),
						("DEC", ("011101", 1)),
						("JMP", ("011110", 1)),
						("CMP", ("011111", 2)),
						("JZ", ("100000", 1)),
						("JN", ("100001", 1)),
						("JO", ("100010", 1)),
						("HLT", ("100011", 0))
					]
				),
				registers: HashMap::from(
					[
						("A", 0),
						("B", 1),
						("C", 2),
						("D", 3),
						("F", 4),
						("PC", 5),
						("SP", 6)
					]
				),
				line: line.to_string(),
				error: error
			}
		}

		pub fn compile(&self) -> Vec<String> {
			let mut binary: Vec<String> = Vec::new();
			let instruction: Vec<&str> = self.line.split(" ").collect();
			
			if self.instructions.contains_key(instruction.get(0).unwrap().to_uppercase().as_str()) {
				let cmd = &self.instructions.get(&instruction.get(0).unwrap().to_uppercase() as &str).unwrap().0.to_string();
				let args = &instruction[1..];

				println!("Current instruction: {} -> {}", instruction.get(0).unwrap().to_lowercase(), cmd);

				binary.push(cmd.to_string());

				if args.len() == self.instructions.get(&instruction.get(0).unwrap().to_uppercase() as &str).unwrap().1 as usize {
					for arg in args {
						let mut i = arg.chars().into_iter();
						let prefix = i.next().unwrap();
						let mut bin = String::new();
						let mut start = 1;
	
						if &arg[1..] == "" {
							self.error.print_stacktrace("RegisterIDError", "Missing argument info".to_string());
						}
						
						if prefix == '!' {
							bin = self.bin(&arg[1..]);
						}
						else if prefix == '@' {
							if !self.registers.contains_key(&arg[1..]) {
								self.error.print_stacktrace("RegisterIDError", format!("Unknown register ID '{}'", &arg[1..]));
							}
	
							// if &(*self.registers.get(&arg[1..].to_uppercase() as &str).unwrap() as usize) <= &4 {
							// 	bin = self.bin(&self.registers.get(&arg[1..].to_uppercase() as &str).unwrap().to_string());
							// }
							// else {
							// 	self.error.print_stacktrace("RegisterIDError".to_string(), format!("Illegal use of register '{}'", &arg[1..]))
							// }
	
							bin = self.bin(&self.registers.get(&arg[1..].to_uppercase() as &str).unwrap().to_string());
						}
						else if prefix == '0' {
							let base = prefix.to_string() + &i.next().unwrap().to_string();
							start = 2;
							
							if base == "0b" {
								bin = arg[2..].to_string();
							}
							else if base == "0x" {
								bin = BinaryString::from_hex(&arg[2..]).unwrap().to_string();
							}
							else {
								self.error.print_stacktrace("BaseError", format!("Unknown base '{}'", base));
							}
						}
						else {
							self.error.print_stacktrace("ArgError", format!("Unknown prefix '{}'", prefix));
						}
	
						bin = bin.trim_start_matches("0").to_string();
						let repeat = cmd.len() + bin.len();
						if repeat > 16 {
							self.error.print_stacktrace("OverflowError", format!("Value '{}' goes over 10-bit limit", &arg[start..]))
						}
	
						binary.push("0".repeat(16 - repeat) + &bin.to_string());
					}
				}
				else {
					if args.len() > self.instructions.get(&instruction.get(0).unwrap().to_uppercase() as &str).unwrap().1 as usize {
						self.error.print_stacktrace("ArgError", format!("Too many arguments; instruction only takes {} argument(s)", self.instructions.get(instruction.get(0).unwrap().to_uppercase().as_str()).unwrap().1));
					}
					else {
						self.error.print_stacktrace("ArgError", format!("Missing {} argument(s)", self.instructions.get(instruction.get(0).unwrap().to_uppercase().as_str()).unwrap().1 - args.len() as u8))
					}
				}
			}
			else {
				self.error.print_stacktrace("InstructionError", format!("Unknown instruction '{}'", instruction.get(0).unwrap()));
			}

			if binary[1..].len() == 2 {
				for arg in &binary.clone()[1..] {
					let mut arg_val = arg.trim_start_matches("0");
					if arg_val == "" {
						arg_val = "0";
					}

					let index = binary.iter().position(|i| i == arg).unwrap();
					binary[index] = "0".repeat(5 - arg_val.len()) + &arg_val.to_string();
				}
			}

			return binary;
		}

		fn bin(&self, immediate: &str) -> String {
			for character in immediate.chars() {
				if !character.is_numeric() {
					self.error.print_stacktrace("ValueError", format!("Invalid value '{}'", immediate));
				}
			}

			let mut bin = 0;
			match immediate.parse::<u16>() {
				Ok(immediate) => bin = immediate,
    			Err(_) => self.error.print_stacktrace("ValueError", format!("Value '{}' goes over 10-bit limit", immediate)),
			}

			return format!("{:b}", bin);
		}
	}
}

File: rks/utils/compiler.py
from .error import Error

class Compiler:
	instructions = {
		"nop": (0, 0),
		"lda": (1, 1),
		"ldb": (2, 1),
		"ldc": (3, 1),
		"ldd": (4, 1),
		"swp": (5, 2),
		"mov": (6, 2),
		"sta": (7, 1),
		"stb": (8, 1),
		"stc": (9, 1),
		"std": (10, 1),
		"lra": (11, 0),
		"lrb": (12, 1),
		"lrc": (13, 1),
		"lrd": (14, 1),
		"psh": (15, 1),
		"pop": (16, 2),
		"isp": (17, 2),
		"dsp": (18, 1),
		"add": (19, 2),
		"sub": (20, 2),
		"mul": (21, 2),
		"div": (22, 2),
		"and": (23, 2),
		"or": (24, 2),
		"nor": (25, 2),
		"xor": (26, 2),
  		"not": (27, 1),
		"inc": (28, 1),
		"dec": (29, 1),
		"cmp": (30, 2),
		"jmp": (31, 1),
		"jz": (32, 1),
		"jnz": (33, 1),
		"jn": (34, 1),
		"jnn": (35, 1),
		"jo": (36, 1),
		"jno": (37, 1),
		"call": (38, 1),
		"ret": (39, 0),
		"out": (40, 1),
		"hlt": (41, 1),
	}
	registers = {
		"a": 0,
		"b": 1,
		"c": 2,
		"d": 3,
		"f": 4,
		"pc": 5,
		"sp": 6,
		"cs": 7
	}

	def __init__(self, instruction: str, labels: dict[str, tuple[int, list[str]]], macros: dict[str, tuple[int, list[str]]], lines: list[str], error: Error) -> None:
		self.instruction = instruction
		self.labels = labels
		self.macros = macros
		self.lines = lines
		self.error = error

	def compile(self) -> list[str]:
		binary = []
		start = 1
		word = ""

		instruction = [i.strip().lower() for i in self.instruction.split(" ")]
		instruction_idx = self.lines.index(self.instruction)
		print(f"Current Instruction: {instruction[0]} -> {'0'*(6-len(bin(self.instructions[instruction[0]][0])[2:])) + bin(self.instructions[instruction[0]][0])[2:]}")

		if len(instruction[1:]) <= self.instructions[instruction[0]][1]:
			binary.append("0"*(6-len(bin(self.instructions[instruction[0]][0])[2:])) + bin(self.instructions[instruction[0]][0])[2:])

			for arg in instruction[1:]:
				argBin = ""
				if arg[0] == "!":
					if arg[1:].isnumeric():
						argBin = "0"*(10-len(bin(int(arg[1:]))[2:])) + bin(int(arg[1:]))[2:]
					else:
						self.error.print_stacktrace("ValueError", f"Invalid immediate '{arg[1:]}'")

				elif arg[0] == "@":
					if arg[1:] in self.registers.keys():
						argBin = "0"*(10-len(bin(self.registers[arg[1:]])[2:])) + bin(self.registers[arg[1:]])[2:]
					else:
						self.error.print_stacktrace("RegisterError", f"Unknown register ID '{arg[1:]}'")

				elif arg[0] == "0":
					base = arg[1]
					start = 0
					if base == "b":
						try:
							temp = int(arg[2:], 2)
						except ValueError:
							self.error.print_stacktrace("ValueError", f"'{arg[2:]}' is not valid binary")
						else:
							argBin = "0"*(10-len(bin(temp)[2:].lstrip("0"))) + bin(temp)[2:].lstrip("0")

					elif base == "x":
						try:
							temp = int(arg[2:].lower(), 16)
						except ValueError:
							self.error.print_stacktrace("ValueError", f"'{arg[2:]}' is not valid hexadecimal")
						else:
							argBin = "0"*(10-len(bin(temp)[2:].lstrip("0"))) + bin(temp)[2:].lstrip("0")

					else:
						self.error.print_stacktrace("ArgError", f"Unknown base '{arg[0:2]}'")

				elif arg[0] == ".":
					try:
						for label in self.labels.keys():
							if self.labels[label][0] > instruction_idx:
								self.labels[label] = (self.labels[label][0]+1, self.labels[label][1],)
        
						word = "0"*(16-len(bin(self.labels[arg[1:]][0])[2:])) + bin(self.labels[arg[1:]][0])[2:]
						argBin = "0"*10

					except KeyError: self.error.print_stacktrace("LabelError", f"Unknown label '{arg[1:]}'")
				
				elif arg[0] == "$":
					try:
						for macro in self.macros.keys():
							if self.macros[macro][0] > instruction_idx:
								self.macros[macro] = (self.labels[macro][0]+1, self.labels[macro][1],)

						word = "0"*(16-len(bin(self.macros[arg[1:]][0])[2:])) + bin(self.macros[arg[1:]][0])[2:]
						argBin = "0"*10

					except KeyError: self.error.print_stacktrace("MacroError", f"Unknown macro '{arg[1:]}'")

				else:
					self.error.print_stacktrace("ArgError", f"Unknown argument prefix '{arg[0]}'")

				if len(argBin) > 10:
					self.error.print_stacktrace("ValueError", f"Argument bit width goes over 10-bit limit: '{arg[start:]}' -> '{argBin}'")

				binary.append(argBin)

			if len(binary[1:]) == 2:
				args = binary[1:]
				for i in range(len(args)):
					args[i] = "0"*(5-len(args[i].lstrip("0"))) + args[i].lstrip("0")

				binary = list(binary[0]) + args

		else:
			extra = ", ".join([f"'{arg}'" for arg in instruction[1:][-(len(instruction[1:]) - self.instructions[instruction[0]][1]):]])
			self.error.print_stacktrace("ArgError", f"Extra arguments {extra}")
   
		if len(binary) == 1:
			binary.append("0"*10)

		return (binary + ["\n" + word]) if word else binary

	@classmethod
	def clean(cls, instructions: list[str]) -> list[str]:
		for i in range(len(instructions)):
			if instructions[i] == "\n" or instructions[i].startswith(";"): instructions[i] = ""
			else: instructions[i] = instructions[i].rstrip()
		return instructions

	@classmethod
	def collect(cls, lines: list[str], location: str) -> tuple[dict[str, tuple[int, list[str]]], dict[str, tuple[int, list[str]]] , list[str]]:
		labels = {}
		macros = {}
		instructions = []

		inLabel = False
		labelName = ""
		inMacro = False
		macroName = ""

		idx = 0
		for i in range(len(lines)):
			if lines[i].startswith("."):
				header = [temp.strip().lower() for temp in lines[i][1:].split(" ")]
				if header[0] == "label":
					labels[header[1]] = (idx, [])
					inLabel = True
					labelName = header[1]

				elif header[0] == "macro":
					macros[header[1]] = (idx, [])
					inMacro = True
					macroName = header[1]

				else: Error(lines[i], i+1, location).print_stacktrace("HeaderError", f"Unknown header type '{header[0]}'")

			elif lines[i] != "":
				if inLabel:
					if lines[i].startswith("    "): labels[labelName][1].append(lines[i].strip())
					else: inLabel = False
				
				elif inMacro:
					if lines[i] != "    ret": macros[macroName][1].append(lines[i].strip())
					else: inMacro = False

				instructions.append(lines[i])
				idx += 1

		return labels, macros, instructions